### ISSUE TITLE
Context validation error on Configure Layout page and empty page output on cancel

### ISSUE THREAD

**Issue by bugfolder** [backdrop/backdrop-issues#5065]
A couple of errors can be encountered when adding/configuring layouts. 

1. Go to Add layout. Pick any name and template.
2. Set the path to `node/%`. Context appears and is set to type Node.
3. Change the path to `foobar/%`. Context switches to "String pass-through".
4. Select context type of Node again from the type selector.
5. Click "Create Layout" to save the layout.
6. Click "Configure Layout" tab to re-configure it. (Note that the context is still set to "Node".)
7. Click "Add visibility condition"
8. Select visibility condition of "Node: Type".
8. Select a node type, e.g., "Page".
9. Click "Add visibility condition".
10. Click "Save Layout".
11.  This produces an error `The condition 'Type is <em class="placeholder">Page</em>' does not have the required contexts at this path. Remove the condition to save the layout.`.
12. Hit the cancel button, and get WSOD at path /system/ajax. Watchdog error is `Notice: Undefined index: #ajax in ajax_form_callback() (line 393 of <snip>/core/includes/ajax.inc).`

The extra HTML in the form error is pretty clearly because the condition summary is getting double-sanitized, so that should be an easy fix.

Don't know why the WSOD is happening, but guessing that it might be related to the question of whether one should be able to put a Node context on a bare % placeholder (whose value could well be an invalid nid) and how rigorously that is enforced.

**Comment by indigoxela** [comment]
I can reproduce it, but have some remarks:

Re step 12. "produces an error" - I'd like to clarify that it's no php error, but a form validation message.

Re step 13 - that's *no* WSOD, but simply an empty page that causes a *php notice*. Path /system/ajax is not supposed to get called directly (GET).

There's a related issue regarding that function in ajax.inc: https://github.com/backdrop/backdrop-issues/issues/5041

**Comment by indigoxela** [comment]
I'll try to more precisely explain what we're trying to fix:

1. Odd value combinations are possible when configuring a layout (not sure how we can address that)
2. Double sanitized string in form validation error message (should be easy)
3. Wrong redirect from that form, if form validation failed in an earlier step

**Comment by bugfolder** [comment]
Just checked to confirm that recent changes to Layouts in 1.x and 1.19.0-preview still produce the issues above. There's a slightly simpler sequence of actions to demonstrate:

1. Go to Add layout. Pick any name and template.
2. Set the path to foobar/%.
3. Set the context type to Node.
4. Click "Create Layout."
5. Click "Configure Layout" tab.
6. Click "Add visibility condition"
7. Select visibility condition of "Node: Type".
8. Select a node type, e.g., "Page".
9. Click "Add visibility condition".
10. Click "Save Layout".
11. This produces a form error: "The condition 'Type is &lt;em class="placeholder"&gt;Page&lt;/em&gt;' does not have the required contexts at this path. Remove the condition to save the layout."
12. Hit the cancel button, and get blank page at path /system/ajax. Watchdog error is "Notice: Undefined index: #ajax in ajax_form_callback() (line 393 of <snip>/core/includes/ajax.inc)."

**Comment by jayelless** [comment]
My investigations showed that this issue does not appear to be caused by double sanitization, but rather by cruft left behind when the form validation function resulted it the reported error message.  Subsequently clicking on the "Cancel" button caused the request to become an ajax response with a call to "/system/ajax". While the submit function associated with the "Cancel button was invoked, it was unable to execute the redirect to the layout listing page because all ajax responses cause the "no_redirect" property of the $form_state array to be set to TRUE.  

If the "Cancel" button was clicked before the "save" and the resulting validation error, then the redirect worked correctly.

I was unable to find out how to prevent the request from being classed as an ajax response.  The fix was therefore to force the "no_redirect" value in the $form_state to always be FALSE so that the required redirect could occur under all circumstances.

**Comment by bugfolder** [comment]
Testing on the sandbox for PR https://github.com/backdrop/backdrop/pull/3698, I can verify that the fix in the PR solves the white-screen error at step 13 of the sequence in the first comment. However, the double-sanitization issue in step 12 is still present:

![foo](https://user-images.githubusercontent.com/29214084/132953835-5dc7941d-e65a-44ad-83d8-5e6c58bc4212.jpg)

**Comment by stpaultim** [comment]
@bugfolder - If you found that the problem still exists, I think we should change the labels back to "Needs Work" and remove "Needs Testing." 

If we know the PR doesn't work in current state, there is no point in additional folks testing it yet, is there?

**Comment by bugfolder** [comment]
Yes.

**Comment by bugfolder** [comment]
Tested, works great! One comment on the code.

**Comment by bugfolder** [comment]
Oops, I was premature with that. Looks like one of the tests is failing.

**Comment by jayelless** [comment]
That was a date test, so no idea why the test failed?  I have re-commited the changes and re-pushed.

**Comment by bugfolder** [comment]
I think it's unrelated. The test ("Date Theme") fails on the PR sandbox. But it also fails on my local site based on the 1.x branch.

**Comment by bugfolder** [comment]
Hah! There's a bug in the test. It will fail for 10 more minutes, and then after 0:00 UTC, it will start passing again, I'll warrant.

**Comment by jayelless** [comment]
OK.  I re-added the t() call and the tests all passed.  I will wait an hour and remove the call again :)

**Comment by quicksketch** [comment]
Thanks @jayelless for fixing this! And extra kudos for @bugfolder for the extensive steps to reproduce. Thanks @indigoxela and @klonos for feedback and reviewing. I merged https://github.com/backdrop/backdrop/pull/3698 into 1.x and 1.20.x.
